assets/src/edit-story/components/checklist/checks/videoOptimization.js

```diff
@@ -0,0 +1,244 @@
+import { PRIORITY_COPY } from '../constants';
+import { LayerThumbnail, Thumbnail, THUMBNAIL_TYPES } from '../../thumbnail';
+import {
+  CARD_TYPE,
+  ChecklistCard,
+  DefaultFooterText,
+} from '../../checklistCard';
+import { filterStoryElements } from '../utils/filterStoryElements';
+import {
+  Button,
+  BUTTON_SIZES,
+  BUTTON_TYPES,
+  Tooltip,
+} from '../../../../design-system';
+import { useHighlights } from '../../../app/highlights';
+import { useRegisterCheck } from '../checkCountContext';
+import { StyledVideoOptimizationIcon } from '../../checklistCard/styles';
+
+const OptimizeButton = styled(Button)`
+  margin-top: 4px;
+`;
+
+export function videoElementsNotOptimized(element = {}) {
+  if (element.resource?.isTranscoding) {
+    return true;
+  }
+
+  if (
+    element.type !== 'video' ||
+    element.resource?.local ||
+    element.resource?.isOptimized
+  ) {
+    return false;
+  }
+
+  const videoArea =
+    (element.resource?.height ?? 0) * (element.resource?.width ?? 0);
+  const isLargeVideo =
+    videoArea >= VIDEO_SIZE_THRESHOLD.WIDTH * VIDEO_SIZE_THRESHOLD.HEIGHT;
+  return isLargeVideo;
+}
+
+const initialState = {};
+const actionTypes = {
+  cancelled: 'cancelled',
+  uploaded: 'uploaded',
+  uploading: 'uploading',
+};
+
+function reducer(state, action) {
+  switch (action.type) {
+    case actionTypes.cancelled:
+      return {
+        ...state,
+        [action.element.resource.id]: actionTypes.cancelled,
+      };
+    case actionTypes.uploaded:
+      return {
+        ...state,
+        [action.element.resource.id]: actionTypes.uploaded,
+      };
+    case actionTypes.uploading:
+      return {
+        ...state,
+        [action.element.resource.id]: actionTypes.uploading,
+      };
+    default:
+      throw new Error();
+  }
+}
+
+export const BulkVideoOptimization = () => {
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const isBulkVideoOptimizationEnabled = useFeature(
+    'enableBulkVideoOptimization'
+  );
+  const story = useStory(({ state: storyState }) => storyState);
+  const unoptimizedVideos = filterStoryElements(
+    story,
+    videoElementsNotOptimized
+  );
+
+  const setHighlights = useHighlights(({ setHighlights }) => setHighlights);
+
+  const { optimizeVideo } = useLocalMedia(({ actions }) => ({
+    optimizeVideo: actions.optimizeVideo,
+  }));
+
+  const processVideoElement = useCallback(
+    async (element) => {
+      if (
+        !element.resource.isTranscoding &&
+        state[element.resource.id] !== actionTypes.uploading
+      ) {
+        dispatch({ type: actionTypes.uploading, element });
+        try {
+          await optimizeVideo({ resource: element.resource });
+        } catch (e) {
+          dispatch({ type: actionTypes.cancelled, element });
+          return;
+        }
+        dispatch({ type: actionTypes.uploaded, element });
+      }
+    },
+    [optimizeVideo, state]
+  );
```

Comment:
I think eventually we should move this state closer to the `useProcessMedia`/`useLocalMedia`/`useMediaUploadQueue` logic. This component here shouldn't really have to worry about it.